### PR TITLE
RSDK-1003 - allow only alphanumeric resource names

### PR DESCRIPTION
**Pull request by bhaney**
Constricts a resources name to only be alphanumeric. Allows underscores, but not spaces, dots, commas, etc.

### CODE REVIEW

File: config/resource.go [hunk 1]
```diff
@@ -136,6 +137,13 @@ func (config *Component) Validate(path string) ([]string, error) {
 	if config.Name == "" {
 		return nil, utils.NewConfigValidationFieldRequiredError(path, "name")
 	}
+	matched, err := regexp.MatchString(`^\w+$`, config.Name)
```
Comment (edaniels) [inline config/resource.go:140]: We should just do a simple character check here. I think underscore is okay but you can only start with a letter.

**Review by edaniels — CHANGES_REQUESTED** [review]
Just a change to make

**Comment by bhaney** [comment]
Since "built-in" is a common service name, we will also need to include dashes as valid

**Comment by bhaney** [comment]
The rules are now 
1. cannot begin with a number, dash, or underscore
2. but numbers, dashes, and underscores are allowed in the body of the name

This seems like a job for regex

**Comment by bhaney** [comment]
If you still feel strongly that it shouldn't be a regex, I will get rid of it

File: config/resource.go [hunk 2]
```diff
@@ -136,6 +137,13 @@ func (config *Component) Validate(path string) ([]string, error) {
 	if config.Name == "" {
 		return nil, utils.NewConfigValidationFieldRequiredError(path, "name")
 	}
+	matched, err := regexp.MatchString(`^[a-zA-Z][-\w]*$`, config.Name)
```
Comment (edaniels) [inline config/resource.go:140]: regex is fine but pull this out to a variable. otherwise it compiles every time

File: config/resource.go [hunk 3]
```diff
@@ -397,6 +405,13 @@ func (config *Service) Validate(path string) ([]string, error) {
 		golog.Global().Debugw("no name given, defaulting name to builtin")
 		config.Name = resource.DefaultServiceName
 	}
+	matched, err := regexp.MatchString(`^[a-zA-Z][-\w]*$`, config.Name)
```
Comment (edaniels) [inline config/resource.go:408]: same here. can pull this out to a common function

**Review by edaniels — CHANGES_REQUESTED** [review]

File: config/resource.go (hunk 2 above)
Comment (bhaney) [reply config/resource.go]: compiled and added a comment to the pattern

File: config/resource.go [hunk 4]
```diff
@@ -136,6 +137,13 @@ func (config *Component) Validate(path string) ([]string, error) {
 	if config.Name == "" {
 		return nil, utils.NewConfigValidationFieldRequiredError(path, "name")
 	}
+	matched, err := regexp.MatchString(`^[a-zA-Z][-\w]*$`, config.Name)
+	if !matched {
+		return nil, errors.Errorf("name %q is not fully alphanumeric", config.Name)
```
[the remark this answers is not in the document]
Comment (bhaney) [reply config/resource.go]: changed the error message

File: config/resource.go (hunk 3 above)
Comment (bhaney) [reply config/resource.go]: done

**Review by edaniels — APPROVED** [review]
LGTM!

**Pull request merged**